### PR TITLE
Fix disparate score token types being used as one type

### LINKED ISSUE #24967: Replay Download Gives Incorrect Replay
### Type

Game behaviour

### Bug description

Was checking if I can download my (lazer) replay since I don't have it locally on this [beatmap's difficulty](https://osu.ppy.sh/beatmapsets/523558#osu/1205766) and prompts me with an error saying I'm missing the beatmap. I download the "missing" beatmap and download the replay again, which downloads someone else's replay on that "missing" beatmap.

Though, I assume this replay may not actually exist as its over a year old and all other beatmaps don't have an online replay available thats 5+ months old. It just happens to allow me to "download" this one.

This issue might've been fixed as mentioned in [discord](https://discord.com/channels/188630481301012481/1097318920991559880/1157208409104465940).

### Screenshots or videos

https://github.com/ppy/osu/assets/68358837/4a620f03-310c-4f95-8be6-aafc7cdd644e

### Version

Latest Release - 2023.924.1

### Logs

[database.log](https://github.com/ppy/osu/files/12762929/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/12762932/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/12762933/network.log)
[performance.log](https://github.com/ppy/osu/files/12762934/performance.log)

### PR DESCRIPTION
**Pull request by bdach**
More stuff related to https://github.com/ppy/osu/issues/24229. I said I was gonna fix single player replays first, but on second thought I just could not look away from the magnitude of this breakage.

---

- Closes https://github.com/ppy/osu/issues/24967

The replay upload flow and the user statistics update flow would rely on the score token sent by client to spectator server in `BeginPlaySession()`, in order to unmap the score token to a solo score ID when processing finishes.

However, spectator server thus far has not been aware that there are actually _two_ types of token - and to be fair, client gave it no practical way to actually know this. In solo, the token ID comes from the `solo_score_tokens` table, and in multiplayer, the token ID comes from the `multiplayer_score_links` table (after https://github.com/ppy/osu-web/pull/10437). Spectator server has however been treating both types as the solo score type. In particular, this could mean that the replay recording flow would royally confuse things and unmap the token to a score ID wrong due to looking for it in the wrong table, and as such, upload an entirely misattributed replay. (See also: [relevant discord conversation](https://discord.com/channels/188630481301012481/188630616286167041/1147069430812127232).)

This PR aims to clean this up by passing a `ScoreToken` structure to spectator server rather than a plain `long`.

I'm not adding a dependency on https://github.com/ppy/osu-web/pull/10437 here because the current token handling is wrong anyhow and would have needed to be fixed. That pull only begins to functionally matter for the upcoming `osu-server-spectator`-side change.

## Deployment

This is a straight-up breaking change, as in:

| | old spectator-server | new spectator-server |
| -: | :-: | :-: |
| **old client** | :green_circle: OK | :red_circle: broken |
| **new client** | :red_circle: broken | :green_circle: OK |

by which I mean that after the upcoming `osu-spectator-server`-side change is deployed, old clients will not be able to spectate.

I wanted to make this not breaking, but gave up because signalr is just not allowing that to happen because [you can't overload hub methods, period](https://stackoverflow.com/questions/68664120/asp-net-core-signalr-hub-method-overloading). So I just did a straight-up API break. For whatever that's worth, [we did that last time, too](https://github.com/ppy/osu/pull/21588).

If you think backcompat is a blocker here, then I'll try again, but it'll 99% take at least a method rename.

### CODE REVIEW

File: osu.Game/Online/ScoreToken.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    /// Represents a score token that can later be exchanged with web for a score ID.
+    /// </summary>
+    /// <remarks>
+    /// The correct way of exchanging the token for a score ID depends on the <see cref="Type"/>.
+    /// </remarks>
+    [Serializable]
+    [MessagePackObject]
+    public record ScoreToken(
+        [property: Key(0)] long ID,
+        [property: Key(1)] ScoreTokenType Type);
```
Comment (bdach) [inline osu.Game/Online/ScoreToken.cs:19]: Not sure you'll like this syntax, especially the `[property: ]` stuff. I did it this way because it's kinda convenient for generating equality members & `.ToString()` which begins to matter on the spectator-server side. But I'll switch to standard `class` if this offends.

File: osu.Game/Screens/Play/SubmittingPlayer.cs [hunk 2]
```diff
@@ -176,13 +178,19 @@ public override bool OnExiting(ScreenExitEvent e)
         [CanBeNull]
         protected abstract APIRequest<APIScoreToken> CreateTokenRequest();
 
+        /// <summary>
+        /// Constructs a <see cref="ScoreToken"/> for later use
+        /// from the <see cref="APIScoreToken"/> returned by web.
+        /// </summary>
+        protected abstract ScoreToken RetrieveScoreToken(APIScoreToken token);
```
Comment (bdach) [inline osu.Game/Screens/Play/SubmittingPlayer.cs:185]: I generally wanted to avoid creating this method and try to constrain this in `CreateTokenRequest()`, but the C# type system makes this annoying.

tl;dr: I would ideally do

```csharp
protected abstract APIRequest<IAPIScoreToken> CreateTokenRequest()
```

with two implementations of `IAPIScoreToken` for solo and for multi. However the C# type system makes this extra annoying, since you can only have variance on interface generics, so that'd mean I'd have to do `IAPIRequest<T>` first, then probably something like `IAPIScoreTokenRequest<T>`, only then can it have `IAPIScoreToken` in the generic...... I don't think so. Unless you consider that better I guess.

**Comment by peppy** [comment]
As I mentioned in discord, sorry about this but I don't think we'll be requiring this (ignoring the wasted effort, that's probably a huge plus). I've been holding off reviewing this while I got my thoughts together and reassessed the whole structure of things, which I've documented over on https://github.com/ppy/osu-infrastructure/issues/24.

Won't close just yet, but going to block until we confirm the direction in the linked issue is feasible (I believe @nanaya has already started osu-web implementation of it).

**Comment by peppy** [comment]
Closing for now to keep things clean. See https://github.com/ppy/osu-infrastructure/issues/24 for rationale.
